### PR TITLE
Entry Array

### PR DESCRIPTION
**Pull request by sanastas**
Adding EntryArray that encapsulates all functionality required to keep array of entries on top of primitive longs array

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: core/src/main/java/com/yahoo/oak/Chunk.java [hunk 1]
```diff
@@ -167,10 +181,10 @@ boolean allocateEntry(ThreadContext ctx, K key) {
     }
 
     /**
-     * See {@code EntrySet.writeValueStart(ThreadContext)} for more information
+     * See {@code EntrySet.allocateValue(ThreadContext)} for more information
      */
     void writeValue(ThreadContext ctx, V value, boolean writeForMove) {
```
Comment (sanastas) [inline core/src/main/java/com/yahoo/oak/Chunk.java:186]: Probably should rename to writeValueStart...

File: core/src/main/java/com/yahoo/oak/Chunk.java [hunk 2]
```diff
@@ -258,7 +272,7 @@ int compareKeyAndEntryIndex(KeyBuffer tempKeyBuff, K key, int ei) {
      *                   In this case, {@code ctx.isKeyValid() == False} and {@code ctx.isValueValid() == False}.
      *             (2) {@code key} was found.
      *                   In this case, {@code (ctx.isKeyValid() == True}
-     *                   The state of the value associated with {@code key} is described in {@code ctx.valueState}.
+     *                   The state of the value associated with {@code key} is described in {@code ctx.entryState}.
      *                   It can be one of the following states:
      *                     (1) not yet inserted, (2) in the process of being inserted, (3) valid,
```
Comment (sanastas) [inline core/src/main/java/com/yahoo/oak/Chunk.java:277]: For OakMap there is no such thing as "(2) in the process of being inserted"

File: core/src/main/java/com/yahoo/oak/Chunk.java [hunk 3]
```diff
@@ -320,27 +334,29 @@ private int binaryFind(KeyBuffer tempKey, K key) {
...
         // return NONE_NEXT to indicate that a regular linear search is needed
-        if (compareKeyAndEntryIndex(tempKey, key, entrySet.getHeadNextIndex()) <= 0) {
+        if (compareKeyAndEntryIndex(tempKey, key, entrySet.getHeadNextEntryIndex()) <= 0) {
             return NONE_NEXT;
         }
 
         // optimization: compare with last key to avoid binary search (here sortedCount is not zero)
-        if (compareKeyAndEntryIndex(tempKey, key, sortedCount) > 0) {
-            return sortedCount;
+        if (compareKeyAndEntryIndex(tempKey, key, sortedCount - 1) > 0) {
```
Comment (sanastas) [inline core/src/main/java/com/yahoo/oak/Chunk.java:342]: probably better to have some getLastEntryIndex() method for (sortedCount - 1) used everywhere...

File: core/src/main/java/com/yahoo/oak/Chunk.java [hunk 4]
```diff
@@ -320,27 +334,29 @@ private int binaryFind(KeyBuffer tempKey, K key) {
...
+            return sortedCount - 1;
         }
 
-        int start = 0;
+        // `start` and `end` are intentionally initiated outside of the array boundaries.
```
Comment (sanastas) [inline core/src/main/java/com/yahoo/oak/Chunk.java:346]: So the returned key as exactly the key we are looking for, or the maximal one still less than the key we are looking for...

File: core/src/main/java/com/yahoo/oak/Chunk.java [hunk 5]
```diff
@@ -414,17 +430,15 @@ int linkEntry(ThreadContext ctx, K key) {
         int prev;
```
Comment (sanastas) [inline core/src/main/java/com/yahoo/oak/Chunk.java:430]: To rename linkEntry()?

File: core/src/main/java/com/yahoo/oak/Chunk.java [hunk 6]
```diff
@@ -582,12 +603,10 @@ final int copyPartNoKeys(ValueBuffer tempValue, Chunk<K, V> srcChunk, final int
         }
```
Comment (sanastas) [inline core/src/main/java/com/yahoo/oak/Chunk.java:603]: Rename copyPartNoKeys() to copyPartOfEntries()

File: core/src/main/java/com/yahoo/oak/EntryArray.java [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+     * @param kMM off-heap allocations and releases for keys
+     * @param entriesCapacity how many entries should this EntrySet keep at maximum
+     * @param keySerializer   used to serialize the key when written to off-heap
+     */
+    EntryArray(MemoryManager vMM, MemoryManager kMM, int additionalFieldCount, int entriesCapacity,
+               OakSerializer<K> keySerializer,
+             OakSerializer<V> valueSerializer) {
+        this.valuesMemoryManager = vMM;
+        this.keysMemoryManager = kMM;
+        this.fields = additionalFieldCount + 2;
```
Comment (sanastas) [inline core/src/main/java/com/yahoo/oak/EntryArray.java:58]: Add comments for this line

File: core/src/main/java/com/yahoo/oak/EntryArray.java [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+     * Atomically reads the key reference from the entry (given by entry index "ei")
+     */
+    protected long getKeyReference(int ei) {
+        return getEntryFieldLong(ei, KEY_REF_OFFSET);
+    }
+
+    /**
+     * Atomically writes the key reference to the entry (given by entry index "ei")
+     */
+    protected void setKeyReference(int ei, long value) {
```
Comment (sanastas) [inline core/src/main/java/com/yahoo/oak/EntryArray.java:171]: Probably to add casKeyReference, as it would be needed for OakHash

File: core/src/main/java/com/yahoo/oak/EntrySet.java [hunk 9]
```diff
@@ -32,83 +30,48 @@
  * -----------------------------------------------------------------------------|
  * 3 | NOT IN USE:             saved for alignment and future usage             |
  * --------------------------------------------------------------------------------------
- * 4 | KEY_REFERENCE          | these 2 integers together, represented as long  | entry with
- * --|                        | provide KEY_REFERENCE. Pay attention that       | entry index
- * 5 |                        | KEY_REFERENCE is different than VALUE_REFERENCE | 1
+ * 4 | KEY_REF_OFFSET          | these 2 integers together, represented as long  | entry with
+ * --|                        | provide KEY_REF_OFFSET. Pay attention that       | entry index
```
Comment (sanastas) [inline core/src/main/java/com/yahoo/oak/EntrySet.java:34]: Change this comment

File: core/src/main/java/com/yahoo/oak/EntrySet.java [hunk 10]
```diff
@@ -32,83 +30,48 @@
...
-        OFFSET(int value) {
-            this.value = value;
-        }
-    }
-
-    static final int INVALID_ENTRY_INDEX = 0;
-
-    // location of the first (head) node - just a next pointer (always same value 0)
-    private final int headNextIndex = 0;
+    private static final int NEXT_FIELD = 2;
```
Comment (sanastas) [inline core/src/main/java/com/yahoo/oak/EntrySet.java:62]: Rename to offset

File: core/src/main/java/com/yahoo/oak/EntrySet.java [hunk 11]
```diff
@@ -287,117 +127,44 @@ int getHeadNextIndex() {
...
     /**
-     * Reads a key from entry at the given entry index (from off-heap).
-     * Returns false if:
-     *   (1) there is no such entry or
-     *   (2) entry has no key set
-     *
-     * @param key the buffer that will contain the key
-     * @param ei  the entry index to read
-     * @return true if the entry index has a valid key allocation reference
+     * TODO: doc
```
Comment (sanastas) [inline core/src/main/java/com/yahoo/oak/EntrySet.java:134]: Resolve TODO

File: core/src/main/java/com/yahoo/oak/EntrySet.java [hunk 12]
```diff
@@ -287,117 +127,44 @@ int getHeadNextIndex() {
...
-
-    /********************************************************************************************/
-    /* Methods for managing the entry context of the keys and values inside ThreadContext       */
-
     /**
-     * Updates the key portion of the entry context inside {@code ctx} that matches its entry context index.
-     * Thus, {@code ctx.initEntryContext(int)} should be called prior to this method on this {@code ctx} instance.
-     *
-     * @param ctx the context that will be updated and follows the operation with this key
+     * TODO: doc
```
Comment (sanastas) [inline core/src/main/java/com/yahoo/oak/EntrySet.java:151]: Resolve TODO

**Review by sanastas — COMMENTED** [review]
To be continued

**Review by sanastas — COMMENTED** [review]
My side of the review is done.

File: core/src/main/java/com/yahoo/oak/Chunk.java (hunk 1 above)
Comment (liran-funaro) [reply core/src/main/java/com/yahoo/oak/Chunk.java:186]: I'm indifferent to the name, but I suggest matching the names between `Chunk.writeValue` to `EntrySet.allocateValue`.
If you rename it to `Chunk.writeValueStart`, then also rename `EntrySet.writeValueStart`.

File: core/src/main/java/com/yahoo/oak/Chunk.java (hunk 3 above)
Comment (liran-funaro) [reply core/src/main/java/com/yahoo/oak/Chunk.java:342]: Did you mean `getLastSortedEntryIndex()`?

File: core/src/main/java/com/yahoo/oak/EntryArray.java (hunk 8 above)
Comment (liran-funaro) [reply core/src/main/java/com/yahoo/oak/EntryArray.java:171]: IMO, it is best to add methods when needed. Otherwise, you wouldn't have a code path to this method and you wouldn't be able to test it.

File: core/src/main/java/com/yahoo/oak/EntryArray.java [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+package com.yahoo.oak;
+
+import sun.misc.Unsafe;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+/*
+* EntryArray is array of entries based on primitive array of longs.
+* Entry must include at least Key Reference and Value Reference fields, but might include more.
+* */
```
Comment (liran-funaro) [inline core/src/main/java/com/yahoo/oak/EntryArray.java:16]: Move the field chart from EntrySet to here and describe the optional additional fields

File: core/src/main/java/com/yahoo/oak/EntrySet.java [hunk 14]
```diff
@@ -32,83 +30,48 @@
  * -----------------------------------------------------------------------------|
  * 3 | NOT IN USE:             saved for alignment and future usage             |
  * --------------------------------------------------------------------------------------
- * 4 | KEY_REFERENCE          | these 2 integers together, represented as long  | entry with
- * --|                        | provide KEY_REFERENCE. Pay attention that       | entry index
- * 5 |                        | KEY_REFERENCE is different than VALUE_REFERENCE | 1
+ * 4 | KEY_REF_OFFSET          | these 2 integers together, represented as long  | entry with
```
Comment (liran-funaro) [inline core/src/main/java/com/yahoo/oak/EntrySet.java:33]: Why the change to "ref_offset"? We call it a reference everywhere else. It implies including all the information to access the key/value. This includes the offset.

File: core/src/main/java/com/yahoo/oak/EntrySet.java (hunk 10 above)
Comment (liran-funaro) [reply core/src/main/java/com/yahoo/oak/EntrySet.java:62]: I think that "entry field" describes it better.
Anyway, the comment is not correct anymore.

**Review by liran-funaro — APPROVED** [review]
LGTM

**Review by sanastas — COMMENTED** [review]
Really last comments :)

File: core/src/main/java/com/yahoo/oak/Chunk.java (hunk 1 above)
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/Chunk.java:186]: all changed to `allocateValue`

File: core/src/main/java/com/yahoo/oak/Chunk.java (hunk 2 above)
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/Chunk.java:277]: fixed

File: core/src/main/java/com/yahoo/oak/Chunk.java (hunk 4 above)
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/Chunk.java:346]: fixed

File: core/src/main/java/com/yahoo/oak/Chunk.java (hunk 5 above)
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/Chunk.java:430]: I will leave this name as is for now...

File: core/src/main/java/com/yahoo/oak/Chunk.java (hunk 6 above)
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/Chunk.java:603]: done

File: core/src/main/java/com/yahoo/oak/EntryArray.java (hunk 7 above)
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/EntryArray.java:58]: fixed

File: core/src/main/java/com/yahoo/oak/EntryArray.java (hunk 8 above)
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/EntryArray.java:171]: OK, skipping this comment

File: core/src/main/java/com/yahoo/oak/EntrySet.java (hunk 9 above)
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/EntrySet.java:34]: fixed

File: core/src/main/java/com/yahoo/oak/EntrySet.java (hunk 10 above)
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/EntrySet.java:62]: This number is used as an offset to next field inside an entry

File: core/src/main/java/com/yahoo/oak/EntrySet.java (hunk 11 above)
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/EntrySet.java:134]: fixed

File: core/src/main/java/com/yahoo/oak/EntrySet.java (hunk 12 above)
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/EntrySet.java:151]: fixed

File: core/src/main/java/com/yahoo/oak/Chunk.java [hunk 15]
```diff
@@ -29,7 +30,6 @@
     private static final double SORTED_REBALANCE_RATIO = 2;
     private static final double MAX_ENTRIES_FACTOR = 2;
     private static final double MAX_IDLE_ENTRIES_FACTOR = 5;
-    private static final int INVALID_ANCHOR_INDEX = -1;
 
     // defaults
     public static final int MAX_ITEMS_DEFAULT = 4096;
```
[the remark this answers is not in the document]
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/Chunk.java:35]: fixed

File: core/src/main/java/com/yahoo/oak/EntryArray.java (hunk 13 above)
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/EntryArray.java:16]: done

File: core/src/main/java/com/yahoo/oak/EntrySet.java (hunk 14 above)
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/EntrySet.java:33]: fixed the comment

File: core/src/main/java/com/yahoo/oak/Chunk.java [hunk 16]
```diff
@@ -117,10 +117,24 @@ private Chunk(int maxItems, AtomicInteger externalSize, MemoryManager vMM, Memor
...
+    private void duplicateKeyBuffer(KeyBuffer src, KeyBuffer dst) {
+        final int keySize = src.capacity();
+        dst.getSlice().allocate(keySize, false);
+
+        // We duplicate the buffer without instantiating a write buffer because the user is not involved.
+        UnsafeUtils.UNSAFE.copyMemory(src.getAddress(), dst.getAddress(), keySize);
+    }
+
     /********************************************************************************************/
     /*-----------------------------  Wrappers for EntrySet methods -----------------------------*/
```
[the remark this answers is not in the document]
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/Chunk.java:140]: fixed

File: core/src/main/java/com/yahoo/oak/Chunk.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (sanastas) [reply core/src/main/java/com/yahoo/oak/Chunk.java:186]: done
